{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Comp 3550 bioinformatics project"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "some stuff"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Libary setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Import necessary libraries\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from Bio.Blast import NCBIWWW, NCBIXML\n",
    "from Bio import SeqIO\n",
    "import csv, os\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions to convert Fasta Files to CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Functions to convert fasta files to csv\n",
    "\n",
    "def convertFastaToCsv(input_file, output_file):\n",
    "    with open(input_file, 'r') as infile, open(output_file, 'w', newline='') as outfile:\n",
    "        csv_writer = csv.writer(outfile)\n",
    "        csv_writer.writerow(['PrimerPair', 'Direction', 'Sequence'])\n",
    "        \n",
    "        for line in infile:\n",
    "            if line.strip() == '':\n",
    "                continue\n",
    "            if line.startswith('>'):\n",
    "                parts = line[1:].strip().split('_')\n",
    "                index = parts[1] if len(parts) > 2 else '1'\n",
    "                directionLetter = parts[2].split(\" \")[0] if len(parts) > 2 else parts[1].split(\" \")[0]\n",
    "                if directionLetter == 'F':\n",
    "                    direction = 'forward'\n",
    "                else:\n",
    "                    direction = 'reverse'\n",
    "            else:\n",
    "                sequence = line.strip()\n",
    "                csv_writer.writerow([index, direction, sequence])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convertFastaFolderToCsv(input_folder, output_folder):\n",
    "    if not os.path.exists(output_folder):\n",
    "        os.makedirs(output_folder)\n",
    "\n",
    "    for filename in os.listdir(input_folder):\n",
    "        if filename.endswith('.fas'):\n",
    "            input_file = os.path.join(input_folder, filename)\n",
    "            output_file = os.path.join(output_folder, f'{filename}.csv')\n",
    "            \n",
    "            # Check if output file already exists\n",
    "            if os.path.exists(output_file):\n",
    "                output_file = os.path.join(output_folder, f'{filename}.csv')\n",
    "                continue\n",
    "            \n",
    "            convertFastaToCsv(input_file, output_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Functions for manipulating DataFrames to have the same shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Manipulating Geneious DataFrames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Group Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reshapeGroupGeneious(dataframe):\n",
    "    df = dataframe\n",
    "    \n",
    "    # Create a mapping of primers to their paired primers\n",
    "    primer_pairs = {}\n",
    "    for index, row in df.iterrows():\n",
    "        primer_pairs[row['Name']] = row['Paired Primer']\n",
    "        primer_pairs[row['Paired Primer']] = row['Name']\n",
    "\n",
    "    # Assign primer pair index\n",
    "    df['PrimerPair'] = None\n",
    "    current_index = 1\n",
    "    for primer in primer_pairs:\n",
    "        if df.loc[df['Name'] == primer, 'PrimerPair'].isnull().all():\n",
    "            df.loc[df['Name'] == primer, 'PrimerPair'] = current_index\n",
    "            df.loc[df['Name'] == primer_pairs[primer], 'PrimerPair'] = current_index\n",
    "            current_index += 1\n",
    "\n",
    "    \n",
    "    df = df.rename(columns={\n",
    "        'Sequence (with extension)': 'Sequence',\n",
    "        })\n",
    "    df = df[['PrimerPair',  'Direction', 'Sequence', 'Length (with extension)', 'Paired Primer','Name', 'Minimum', 'Maximum', '# Local Off-target Sites', '%GC', 'Hairpin Tm', 'Self Dimer Tm', 'Tm']]\n",
    "    return df\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Species Dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Manipulating PrimerQuest DataFrames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Group Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reshapeGroupPrimerQuest(dataframe):\n",
    "    df = dataframe\n",
    "    # Extract primer pair number from AssaySet\n",
    "    df['PrimerPair'] = df['AssaySet'].str.extract(r'Assay Set (\\d+)').astype(int)\n",
    "\n",
    "    # Rename Type to Direction and convert to lowercase\n",
    "    df = df.rename(columns={'Type': 'Direction'})\n",
    "    df['Direction'] = df['Direction'].str.replace(' Primer', '').str.lower()\n",
    "\n",
    "    # Drop the old AssaySet column\n",
    "    df = df.drop(columns=['AssaySet'])\n",
    "\n",
    "    # Reorder columns\n",
    "    df = df[['PrimerPair', 'Direction', 'Sequence', 'Start', 'Length', 'Tm', 'GC Percent', 'Amplicon']]\n",
    "    \n",
    "    return df\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Species Dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Manipulating PrimerBlast DataFrames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Group Dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def reshapeGroupPrimerBlast(dataframe):\n",
    "    \n",
    "\n",
    "    # Assuming your DataFrame is named df\n",
    "    df = dataframe\n",
    "\n",
    "    # Melt the DataFrame to long format\n",
    "    df_long = pd.melt(df, id_vars=['Primer pair #', 'Product length'], \n",
    "                    value_vars=[col for col in df.columns if col.startswith('Forward') or col.startswith('Reverse')],\n",
    "                    var_name='Attribute', value_name='Value')\n",
    "\n",
    "    # Extract primer direction and attribute name\n",
    "    df_long['Direction'] = df_long['Attribute'].apply(lambda x: 'Forward' if 'Forward' in x else 'Reverse')\n",
    "    df_long['Attribute'] = df_long['Attribute'].apply(lambda x: x.replace('Forward primer ', '').replace('Reverse primer ', ''))\n",
    "\n",
    "    # Pivot the DataFrame back to wide format\n",
    "    df_wide = df_long.pivot_table(index=['Primer pair #', 'Product length', 'Direction'], \n",
    "                                columns='Attribute', values='Value', aggfunc='first').reset_index()\n",
    "\n",
    "    # Reorder columns\n",
    "    df_wide = df_wide[['Primer pair #', 'Direction', 'Sequence (5\\'->3\\')', 'Template strand', 'Length', 'Start', 'Stop', 'Tm', 'GC%', 'Self complementarity', 'Self 3\\' complementarity', 'Product length']]\n",
    "\n",
    "    df_wide = df_wide.rename(columns={\n",
    "        'Sequence (5\\'->3\\')': 'Sequence',\n",
    "        'Primer pair #': 'PrimerPair'\n",
    "    })\n",
    "    \n",
    "    df_wide['Direction'] = df_wide['Direction'].str.lower()\n",
    "\n",
    "    return df_wide"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Species Dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Manipulating Primer3 DataFrames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Group Dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Species Dataframe"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Function to make Blast Requests"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Main Program"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Setup\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Convert to CSV Files for use with Pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "foldersToConvert = ['data/group/primer3','data/species/primer3']\n",
    "\n",
    "for folder in foldersToConvert:\n",
    "    convertFastaFolderToCsv(folder, folder)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Initialize Group DataFrames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Setup dataframes\n",
    "groupGeneous = pd.read_csv('data/group/geneious/Entero grp-specific primers (GENEIOUS).csv')\n",
    "groupPrimer3 = pd.read_csv('data/group/primer3/EnterobactericaeaP3.fas.csv')\n",
    "groupPrimerBlast = pd.read_csv('data/group/primerBlast/Entero grp-specific primers (BLAST).csv')\n",
    "groupPrimerQuest = pd.read_csv('Data/group/primerQuest/PrimerQuest family-specific.csv')\n",
    "\n",
    "# Store dataframes in dictionary for dynamic usage and access later\n",
    "groupData = {\n",
    "    'Geneious': groupGeneous,\n",
    "    'Primer3': groupPrimer3,\n",
    "    'PrimerBlast': groupPrimerBlast,\n",
    "    'PrimerQuest': groupPrimerQuest\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Investigate shape of data frames\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Minimum</th>\n",
       "      <th>Maximum</th>\n",
       "      <th>Direction</th>\n",
       "      <th>Sequence (with extension)</th>\n",
       "      <th>Length (with extension)</th>\n",
       "      <th>Paired Primer</th>\n",
       "      <th># Local Off-target Sites</th>\n",
       "      <th>%GC</th>\n",
       "      <th>Hairpin Tm</th>\n",
       "      <th>Self Dimer Tm</th>\n",
       "      <th>Tm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>934 F</td>\n",
       "      <td>934</td>\n",
       "      <td>953</td>\n",
       "      <td>forward</td>\n",
       "      <td>GGGCTTGACATACACCGGAA</td>\n",
       "      <td>20</td>\n",
       "      <td>1,130 R</td>\n",
       "      <td>0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.5</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>916 F</td>\n",
       "      <td>916</td>\n",
       "      <td>935</td>\n",
       "      <td>forward</td>\n",
       "      <td>ACGCGAAGAACCTTACCTGG</td>\n",
       "      <td>20</td>\n",
       "      <td>1,131 R</td>\n",
       "      <td>0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.9</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>427 F</td>\n",
       "      <td>427</td>\n",
       "      <td>446</td>\n",
       "      <td>forward</td>\n",
       "      <td>ACGGTACCTGCAGAAGAAGC</td>\n",
       "      <td>20</td>\n",
       "      <td>953 R</td>\n",
       "      <td>0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.4</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>275 F</td>\n",
       "      <td>275</td>\n",
       "      <td>294</td>\n",
       "      <td>forward</td>\n",
       "      <td>CCACACTGGGACTGAGACAC</td>\n",
       "      <td>20</td>\n",
       "      <td>1,182 R</td>\n",
       "      <td>0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>32.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1,182 R</td>\n",
       "      <td>1163</td>\n",
       "      <td>1182</td>\n",
       "      <td>reverse</td>\n",
       "      <td>TAGCATGTGTGAAGCCCTGG</td>\n",
       "      <td>20</td>\n",
       "      <td>275 F</td>\n",
       "      <td>0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Name  Minimum  Maximum Direction Sequence (with extension)  \\\n",
       "0    934 F      934      953   forward      GGGCTTGACATACACCGGAA   \n",
       "1    916 F      916      935   forward      ACGCGAAGAACCTTACCTGG   \n",
       "2    427 F      427      446   forward      ACGGTACCTGCAGAAGAAGC   \n",
       "3    275 F      275      294   forward      CCACACTGGGACTGAGACAC   \n",
       "4  1,182 R     1163     1182   reverse      TAGCATGTGTGAAGCCCTGG   \n",
       "\n",
       "   Length (with extension) Paired Primer  # Local Off-target Sites   %GC  \\\n",
       "0                       20       1,130 R                         0  55.0   \n",
       "1                       20       1,131 R                         0  55.0   \n",
       "2                       20         953 R                         0  55.0   \n",
       "3                       20       1,182 R                         0  60.0   \n",
       "4                       20         275 F                         0  55.0   \n",
       "\n",
       "   Hairpin Tm  Self Dimer Tm    Tm  \n",
       "0         NaN            9.5  60.0  \n",
       "1         NaN           15.9  60.0  \n",
       "2         NaN           17.4  60.0  \n",
       "3        32.7            NaN  60.0  \n",
       "4         NaN            NaN  60.0  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groupData['Geneious'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PrimerPair</th>\n",
       "      <th>Direction</th>\n",
       "      <th>Sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>forward</td>\n",
       "      <td>TGCCTGATGGAGGGGGATAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>reverse</td>\n",
       "      <td>GAGACTCAAGCCTGCCAGTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>forward</td>\n",
       "      <td>GGAACTGAGACACGGTCCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>reverse</td>\n",
       "      <td>TTTAACCTTGCGGCCGTACT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>3</td>\n",
       "      <td>forward</td>\n",
       "      <td>TGGTAGTCCACGCCGTAAAC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   PrimerPair Direction              Sequence\n",
       "0           1   forward  TGCCTGATGGAGGGGGATAA\n",
       "1           1   reverse  GAGACTCAAGCCTGCCAGTT\n",
       "2           2   forward  GGAACTGAGACACGGTCCAG\n",
       "3           2   reverse  TTTAACCTTGCGGCCGTACT\n",
       "4           3   forward  TGGTAGTCCACGCCGTAAAC"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groupData['Primer3'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Primer pair #</th>\n",
       "      <th>Forward primer Sequence (5'-&gt;3')</th>\n",
       "      <th>Forward primer Template strand</th>\n",
       "      <th>Forward primer Length</th>\n",
       "      <th>Forward primer Start</th>\n",
       "      <th>Forward primer Stop</th>\n",
       "      <th>Forward primer Tm</th>\n",
       "      <th>Forward primer GC%</th>\n",
       "      <th>Forward primer Self complementarity</th>\n",
       "      <th>Forward primer Self 3' complementarity</th>\n",
       "      <th>Reverse primer Sequence (5'-&gt;3')</th>\n",
       "      <th>Reverse primer Template strand</th>\n",
       "      <th>Reverse primer Length</th>\n",
       "      <th>Reverse primer Start</th>\n",
       "      <th>Reverse primer Stop</th>\n",
       "      <th>Reverse primer Tm</th>\n",
       "      <th>Reverse primer GC%</th>\n",
       "      <th>Reverse primer Self complementarity</th>\n",
       "      <th>Reverse primer Self 3' complementarity</th>\n",
       "      <th>Product length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>AAGAAGCACCGGCTAACTCC</td>\n",
       "      <td>Plus</td>\n",
       "      <td>20.0</td>\n",
       "      <td>496.0</td>\n",
       "      <td>515.0</td>\n",
       "      <td>60.04</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TTCACAACACGAGCTGACGA</td>\n",
       "      <td>Minus</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1082.0</td>\n",
       "      <td>1063.0</td>\n",
       "      <td>59.90</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>587.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>TATTGCACAATGGGCGCAAG</td>\n",
       "      <td>Plus</td>\n",
       "      <td>20.0</td>\n",
       "      <td>366.0</td>\n",
       "      <td>385.0</td>\n",
       "      <td>59.83</td>\n",
       "      <td>50.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>GGAGTTAGCCGGTGCTTCTT</td>\n",
       "      <td>Minus</td>\n",
       "      <td>20.0</td>\n",
       "      <td>515.0</td>\n",
       "      <td>496.0</td>\n",
       "      <td>60.04</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>150.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>CGGTAATACGGAGGGTGCAA</td>\n",
       "      <td>Plus</td>\n",
       "      <td>20.0</td>\n",
       "      <td>529.0</td>\n",
       "      <td>548.0</td>\n",
       "      <td>59.82</td>\n",
       "      <td>55.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>GTTCTTCGCGTTGCATCGAA</td>\n",
       "      <td>Minus</td>\n",
       "      <td>20.0</td>\n",
       "      <td>980.0</td>\n",
       "      <td>961.0</td>\n",
       "      <td>59.84</td>\n",
       "      <td>50.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>452.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>TTCGATGCAACGCGAAGAAC</td>\n",
       "      <td>Plus</td>\n",
       "      <td>20.0</td>\n",
       "      <td>961.0</td>\n",
       "      <td>980.0</td>\n",
       "      <td>59.84</td>\n",
       "      <td>50.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TTGCGGGACTTAACCCAACA</td>\n",
       "      <td>Minus</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1103.0</td>\n",
       "      <td>1084.0</td>\n",
       "      <td>59.82</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>143.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>TCGTCAGCTCGTGTTGTGAA</td>\n",
       "      <td>Plus</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1063.0</td>\n",
       "      <td>1082.0</td>\n",
       "      <td>59.90</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>TTGTAGCACGTGTGTAGCCC</td>\n",
       "      <td>Minus</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1240.0</td>\n",
       "      <td>1221.0</td>\n",
       "      <td>60.32</td>\n",
       "      <td>55.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>178.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Primer pair # Forward primer Sequence (5'->3')  \\\n",
       "0             1             AAGAAGCACCGGCTAACTCC   \n",
       "1             2             TATTGCACAATGGGCGCAAG   \n",
       "2             3             CGGTAATACGGAGGGTGCAA   \n",
       "3             4             TTCGATGCAACGCGAAGAAC   \n",
       "4             5             TCGTCAGCTCGTGTTGTGAA   \n",
       "\n",
       "  Forward primer Template strand  Forward primer Length  Forward primer Start  \\\n",
       "0                           Plus                   20.0                 496.0   \n",
       "1                           Plus                   20.0                 366.0   \n",
       "2                           Plus                   20.0                 529.0   \n",
       "3                           Plus                   20.0                 961.0   \n",
       "4                           Plus                   20.0                1063.0   \n",
       "\n",
       "   Forward primer Stop  Forward primer Tm  Forward primer GC%  \\\n",
       "0                515.0              60.04                55.0   \n",
       "1                385.0              59.83                50.0   \n",
       "2                548.0              59.82                55.0   \n",
       "3                980.0              59.84                50.0   \n",
       "4               1082.0              59.90                50.0   \n",
       "\n",
       "   Forward primer Self complementarity  \\\n",
       "0                                  4.0   \n",
       "1                                  6.0   \n",
       "2                                  5.0   \n",
       "3                                  6.0   \n",
       "4                                  4.0   \n",
       "\n",
       "   Forward primer Self 3' complementarity Reverse primer Sequence (5'->3')  \\\n",
       "0                                     0.0             TTCACAACACGAGCTGACGA   \n",
       "1                                     2.0             GGAGTTAGCCGGTGCTTCTT   \n",
       "2                                     2.0             GTTCTTCGCGTTGCATCGAA   \n",
       "3                                     0.0             TTGCGGGACTTAACCCAACA   \n",
       "4                                     1.0             TTGTAGCACGTGTGTAGCCC   \n",
       "\n",
       "  Reverse primer Template strand  Reverse primer Length  Reverse primer Start  \\\n",
       "0                          Minus                   20.0                1082.0   \n",
       "1                          Minus                   20.0                 515.0   \n",
       "2                          Minus                   20.0                 980.0   \n",
       "3                          Minus                   20.0                1103.0   \n",
       "4                          Minus                   20.0                1240.0   \n",
       "\n",
       "   Reverse primer Stop  Reverse primer Tm  Reverse primer GC%  \\\n",
       "0               1063.0              59.90                50.0   \n",
       "1                496.0              60.04                55.0   \n",
       "2                961.0              59.84                50.0   \n",
       "3               1084.0              59.82                50.0   \n",
       "4               1221.0              60.32                55.0   \n",
       "\n",
       "   Reverse primer Self complementarity  \\\n",
       "0                                  4.0   \n",
       "1                                  4.0   \n",
       "2                                  6.0   \n",
       "3                                  4.0   \n",
       "4                                  6.0   \n",
       "\n",
       "   Reverse primer Self 3' complementarity  Product length  \n",
       "0                                     2.0           587.0  \n",
       "1                                     0.0           150.0  \n",
       "2                                     2.0           452.0  \n",
       "3                                     2.0           143.0  \n",
       "4                                     0.0           178.0  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groupData['PrimerBlast'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AssaySet</th>\n",
       "      <th>Type</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Start</th>\n",
       "      <th>Length</th>\n",
       "      <th>Tm</th>\n",
       "      <th>GC Percent</th>\n",
       "      <th>Amplicon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Assay Set 1 (Consensus)</td>\n",
       "      <td>Forward Primer</td>\n",
       "      <td>GATGTGCCCAGATGGGATTAG</td>\n",
       "      <td>220.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>62.233</td>\n",
       "      <td>52.381</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Assay Set 1 (Consensus)</td>\n",
       "      <td>Reverse Primer</td>\n",
       "      <td>GTGGATGTCAAGACCAGGTAAG</td>\n",
       "      <td>994.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>62.205</td>\n",
       "      <td>50.000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Assay Set 1 (Consensus)</td>\n",
       "      <td>Product</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>775.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Assay Set 2 (Consensus)</td>\n",
       "      <td>Forward Primer</td>\n",
       "      <td>ACGGTAGCTAATACCGCATAAC</td>\n",
       "      <td>155.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>61.972</td>\n",
       "      <td>45.455</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Assay Set 2 (Consensus)</td>\n",
       "      <td>Reverse Primer</td>\n",
       "      <td>CCTCCAGATCTCTACGCATTTC</td>\n",
       "      <td>707.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>62.034</td>\n",
       "      <td>50.000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>994</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>995</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>996</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>997</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>998</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>999 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    AssaySet            Type                Sequence  Start  \\\n",
       "0    Assay Set 1 (Consensus)  Forward Primer   GATGTGCCCAGATGGGATTAG  220.0   \n",
       "1    Assay Set 1 (Consensus)  Reverse Primer  GTGGATGTCAAGACCAGGTAAG  994.0   \n",
       "2    Assay Set 1 (Consensus)         Product                     NaN    NaN   \n",
       "3    Assay Set 2 (Consensus)  Forward Primer  ACGGTAGCTAATACCGCATAAC  155.0   \n",
       "4    Assay Set 2 (Consensus)  Reverse Primer  CCTCCAGATCTCTACGCATTTC  707.0   \n",
       "..                       ...             ...                     ...    ...   \n",
       "994                      NaN             NaN                     NaN    NaN   \n",
       "995                      NaN             NaN                     NaN    NaN   \n",
       "996                      NaN             NaN                     NaN    NaN   \n",
       "997                      NaN             NaN                     NaN    NaN   \n",
       "998                      NaN             NaN                     NaN    NaN   \n",
       "\n",
       "     Length      Tm  GC Percent  Amplicon  \n",
       "0      21.0  62.233      52.381       NaN  \n",
       "1      22.0  62.205      50.000       NaN  \n",
       "2       NaN     NaN         NaN     775.0  \n",
       "3      22.0  61.972      45.455       NaN  \n",
       "4      22.0  62.034      50.000       NaN  \n",
       "..      ...     ...         ...       ...  \n",
       "994     NaN     NaN         NaN       NaN  \n",
       "995     NaN     NaN         NaN       NaN  \n",
       "996     NaN     NaN         NaN       NaN  \n",
       "997     NaN     NaN         NaN       NaN  \n",
       "998     NaN     NaN         NaN       NaN  \n",
       "\n",
       "[999 rows x 8 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groupData['PrimerQuest']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Investigate dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 10 entries, 0 to 9\n",
      "Data columns (total 12 columns):\n",
      " #   Column                     Non-Null Count  Dtype  \n",
      "---  ------                     --------------  -----  \n",
      " 0   Name                       10 non-null     object \n",
      " 1   Minimum                    10 non-null     int64  \n",
      " 2   Maximum                    10 non-null     int64  \n",
      " 3   Direction                  10 non-null     object \n",
      " 4   Sequence (with extension)  10 non-null     object \n",
      " 5   Length (with extension)    10 non-null     int64  \n",
      " 6   Paired Primer              10 non-null     object \n",
      " 7   # Local Off-target Sites   10 non-null     int64  \n",
      " 8   %GC                        10 non-null     float64\n",
      " 9   Hairpin Tm                 3 non-null      float64\n",
      " 10  Self Dimer Tm              3 non-null      float64\n",
      " 11  Tm                         10 non-null     float64\n",
      "dtypes: float64(4), int64(4), object(4)\n",
      "memory usage: 1.1+ KB\n"
     ]
    }
   ],
   "source": [
    "groupData['Geneious'].info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 20 entries, 0 to 19\n",
      "Data columns (total 3 columns):\n",
      " #   Column      Non-Null Count  Dtype \n",
      "---  ------      --------------  ----- \n",
      " 0   PrimerPair  20 non-null     int64 \n",
      " 1   Direction   20 non-null     object\n",
      " 2   Sequence    20 non-null     object\n",
      "dtypes: int64(1), object(2)\n",
      "memory usage: 612.0+ bytes\n"
     ]
    }
   ],
   "source": [
    "groupData['Primer3'].info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 12 entries, 0 to 11\n",
      "Data columns (total 20 columns):\n",
      " #   Column                                  Non-Null Count  Dtype  \n",
      "---  ------                                  --------------  -----  \n",
      " 0   Primer pair #                           11 non-null     object \n",
      " 1   Forward primer Sequence (5'->3')        10 non-null     object \n",
      " 2   Forward primer Template strand          10 non-null     object \n",
      " 3   Forward primer Length                   10 non-null     float64\n",
      " 4   Forward primer Start                    10 non-null     float64\n",
      " 5   Forward primer Stop                     10 non-null     float64\n",
      " 6   Forward primer Tm                       10 non-null     float64\n",
      " 7   Forward primer GC%                      10 non-null     float64\n",
      " 8   Forward primer Self complementarity     10 non-null     float64\n",
      " 9   Forward primer Self 3' complementarity  10 non-null     float64\n",
      " 10  Reverse primer Sequence (5'->3')        10 non-null     object \n",
      " 11  Reverse primer Template strand          10 non-null     object \n",
      " 12  Reverse primer Length                   10 non-null     float64\n",
      " 13  Reverse primer Start                    10 non-null     float64\n",
      " 14  Reverse primer Stop                     10 non-null     float64\n",
      " 15  Reverse primer Tm                       10 non-null     float64\n",
      " 16  Reverse primer GC%                      10 non-null     float64\n",
      " 17  Reverse primer Self complementarity     10 non-null     float64\n",
      " 18  Reverse primer Self 3' complementarity  10 non-null     float64\n",
      " 19  Product length                          10 non-null     float64\n",
      "dtypes: float64(15), object(5)\n",
      "memory usage: 2.0+ KB\n"
     ]
    }
   ],
   "source": [
    "groupData['PrimerBlast'].info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 999 entries, 0 to 998\n",
      "Data columns (total 8 columns):\n",
      " #   Column      Non-Null Count  Dtype  \n",
      "---  ------      --------------  -----  \n",
      " 0   AssaySet    15 non-null     object \n",
      " 1   Type        15 non-null     object \n",
      " 2   Sequence    10 non-null     object \n",
      " 3   Start       10 non-null     float64\n",
      " 4   Length      10 non-null     float64\n",
      " 5   Tm          10 non-null     float64\n",
      " 6   GC Percent  10 non-null     float64\n",
      " 7   Amplicon    5 non-null      float64\n",
      "dtypes: float64(5), object(3)\n",
      "memory usage: 62.6+ KB\n"
     ]
    }
   ],
   "source": [
    "groupData['PrimerQuest'].info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Clean up dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 10 entries, 0 to 9\n",
      "Data columns (total 21 columns):\n",
      " #   Column                                  Non-Null Count  Dtype  \n",
      "---  ------                                  --------------  -----  \n",
      " 0   index                                   10 non-null     int64  \n",
      " 1   Primer pair #                           10 non-null     object \n",
      " 2   Forward primer Sequence (5'->3')        10 non-null     object \n",
      " 3   Forward primer Template strand          10 non-null     object \n",
      " 4   Forward primer Length                   10 non-null     float64\n",
      " 5   Forward primer Start                    10 non-null     float64\n",
      " 6   Forward primer Stop                     10 non-null     float64\n",
      " 7   Forward primer Tm                       10 non-null     float64\n",
      " 8   Forward primer GC%                      10 non-null     float64\n",
      " 9   Forward primer Self complementarity     10 non-null     float64\n",
      " 10  Forward primer Self 3' complementarity  10 non-null     float64\n",
      " 11  Reverse primer Sequence (5'->3')        10 non-null     object \n",
      " 12  Reverse primer Template strand          10 non-null     object \n",
      " 13  Reverse primer Length                   10 non-null     float64\n",
      " 14  Reverse primer Start                    10 non-null     float64\n",
      " 15  Reverse primer Stop                     10 non-null     float64\n",
      " 16  Reverse primer Tm                       10 non-null     float64\n",
      " 17  Reverse primer GC%                      10 non-null     float64\n",
      " 18  Reverse primer Self complementarity     10 non-null     float64\n",
      " 19  Reverse primer Self 3' complementarity  10 non-null     float64\n",
      " 20  Product length                          10 non-null     float64\n",
      "dtypes: float64(15), int64(1), object(5)\n",
      "memory usage: 1.8+ KB\n"
     ]
    }
   ],
   "source": [
    "groupData['PrimerBlast'].dropna(inplace=True, subset=['Forward primer Template strand'])\n",
    "groupData['PrimerBlast'].reset_index().info()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 10 entries, 0 to 9\n",
      "Data columns (total 9 columns):\n",
      " #   Column      Non-Null Count  Dtype  \n",
      "---  ------      --------------  -----  \n",
      " 0   index       10 non-null     int64  \n",
      " 1   AssaySet    10 non-null     object \n",
      " 2   Type        10 non-null     object \n",
      " 3   Sequence    10 non-null     object \n",
      " 4   Start       10 non-null     float64\n",
      " 5   Length      10 non-null     float64\n",
      " 6   Tm          10 non-null     float64\n",
      " 7   GC Percent  10 non-null     float64\n",
      " 8   Amplicon    0 non-null      float64\n",
      "dtypes: float64(5), int64(1), object(3)\n",
      "memory usage: 852.0+ bytes\n"
     ]
    }
   ],
   "source": [
    "groupData['PrimerQuest'].dropna(inplace=True, subset=['Sequence'])\n",
    "groupData['PrimerQuest'].reset_index().info()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Data Manipulation to have the same shape/column names "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Attribute</th>\n",
       "      <th>PrimerPair</th>\n",
       "      <th>Direction</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Template strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Tm</th>\n",
       "      <th>GC%</th>\n",
       "      <th>Self complementarity</th>\n",
       "      <th>Self 3' complementarity</th>\n",
       "      <th>Product length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>forward</td>\n",
       "      <td>AAGAAGCACCGGCTAACTCC</td>\n",
       "      <td>Plus</td>\n",
       "      <td>20.0</td>\n",
       "      <td>496.0</td>\n",
       "      <td>515.0</td>\n",
       "      <td>60.04</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>587.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>reverse</td>\n",
       "      <td>TTCACAACACGAGCTGACGA</td>\n",
       "      <td>Minus</td>\n",
       "      <td>20.0</td>\n",
       "      <td>1082.0</td>\n",
       "      <td>1063.0</td>\n",
       "      <td>59.9</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>587.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10</td>\n",
       "      <td>forward</td>\n",
       "      <td>GCAGAAGAAGCACCGGCTAA</td>\n",
       "      <td>Plus</td>\n",
       "      <td>20.0</td>\n",
       "      <td>492.0</td>\n",
       "      <td>511.0</td>\n",
       "      <td>60.67</td>\n",
       "      <td>55.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>89.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10</td>\n",
       "      <td>reverse</td>\n",
       "      <td>TGCGCTTTACGCCCAGTAAT</td>\n",
       "      <td>Minus</td>\n",
       "      <td>20.0</td>\n",
       "      <td>580.0</td>\n",
       "      <td>561.0</td>\n",
       "      <td>60.39</td>\n",
       "      <td>50.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>89.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>forward</td>\n",
       "      <td>TATTGCACAATGGGCGCAAG</td>\n",
       "      <td>Plus</td>\n",
       "      <td>20.0</td>\n",
       "      <td>366.0</td>\n",
       "      <td>385.0</td>\n",
       "      <td>59.83</td>\n",
       "      <td>50.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>150.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Attribute PrimerPair Direction              Sequence Template strand Length  \\\n",
       "0                  1   forward  AAGAAGCACCGGCTAACTCC            Plus   20.0   \n",
       "1                  1   reverse  TTCACAACACGAGCTGACGA           Minus   20.0   \n",
       "2                 10   forward  GCAGAAGAAGCACCGGCTAA            Plus   20.0   \n",
       "3                 10   reverse  TGCGCTTTACGCCCAGTAAT           Minus   20.0   \n",
       "4                  2   forward  TATTGCACAATGGGCGCAAG            Plus   20.0   \n",
       "\n",
       "Attribute   Start    Stop     Tm   GC% Self complementarity  \\\n",
       "0           496.0   515.0  60.04  55.0                  4.0   \n",
       "1          1082.0  1063.0   59.9  50.0                  4.0   \n",
       "2           492.0   511.0  60.67  55.0                  4.0   \n",
       "3           580.0   561.0  60.39  50.0                  5.0   \n",
       "4           366.0   385.0  59.83  50.0                  6.0   \n",
       "\n",
       "Attribute Self 3' complementarity  Product length  \n",
       "0                             0.0           587.0  \n",
       "1                             2.0           587.0  \n",
       "2                             1.0            89.0  \n",
       "3                             2.0            89.0  \n",
       "4                             2.0           150.0  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groupData['PrimerBlast'] = reshapeGroupPrimerBlast(groupData['PrimerBlast'])\n",
    "groupData['PrimerBlast'].head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PrimerPair</th>\n",
       "      <th>Direction</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Length (with extension)</th>\n",
       "      <th>Paired Primer</th>\n",
       "      <th>Name</th>\n",
       "      <th>Minimum</th>\n",
       "      <th>Maximum</th>\n",
       "      <th># Local Off-target Sites</th>\n",
       "      <th>%GC</th>\n",
       "      <th>Hairpin Tm</th>\n",
       "      <th>Self Dimer Tm</th>\n",
       "      <th>Tm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>forward</td>\n",
       "      <td>GGGCTTGACATACACCGGAA</td>\n",
       "      <td>20</td>\n",
       "      <td>1,130 R</td>\n",
       "      <td>934 F</td>\n",
       "      <td>934</td>\n",
       "      <td>953</td>\n",
       "      <td>0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.5</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>forward</td>\n",
       "      <td>ACGCGAAGAACCTTACCTGG</td>\n",
       "      <td>20</td>\n",
       "      <td>1,131 R</td>\n",
       "      <td>916 F</td>\n",
       "      <td>916</td>\n",
       "      <td>935</td>\n",
       "      <td>0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.9</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>forward</td>\n",
       "      <td>ACGGTACCTGCAGAAGAAGC</td>\n",
       "      <td>20</td>\n",
       "      <td>953 R</td>\n",
       "      <td>427 F</td>\n",
       "      <td>427</td>\n",
       "      <td>446</td>\n",
       "      <td>0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17.4</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>forward</td>\n",
       "      <td>CCACACTGGGACTGAGACAC</td>\n",
       "      <td>20</td>\n",
       "      <td>1,182 R</td>\n",
       "      <td>275 F</td>\n",
       "      <td>275</td>\n",
       "      <td>294</td>\n",
       "      <td>0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>32.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>reverse</td>\n",
       "      <td>TAGCATGTGTGAAGCCCTGG</td>\n",
       "      <td>20</td>\n",
       "      <td>275 F</td>\n",
       "      <td>1,182 R</td>\n",
       "      <td>1163</td>\n",
       "      <td>1182</td>\n",
       "      <td>0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>60.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  PrimerPair Direction              Sequence  Length (with extension)  \\\n",
       "0          1   forward  GGGCTTGACATACACCGGAA                       20   \n",
       "1          2   forward  ACGCGAAGAACCTTACCTGG                       20   \n",
       "2          3   forward  ACGGTACCTGCAGAAGAAGC                       20   \n",
       "3          4   forward  CCACACTGGGACTGAGACAC                       20   \n",
       "4          4   reverse  TAGCATGTGTGAAGCCCTGG                       20   \n",
       "\n",
       "  Paired Primer     Name  Minimum  Maximum  # Local Off-target Sites   %GC  \\\n",
       "0       1,130 R    934 F      934      953                         0  55.0   \n",
       "1       1,131 R    916 F      916      935                         0  55.0   \n",
       "2         953 R    427 F      427      446                         0  55.0   \n",
       "3       1,182 R    275 F      275      294                         0  60.0   \n",
       "4         275 F  1,182 R     1163     1182                         0  55.0   \n",
       "\n",
       "   Hairpin Tm  Self Dimer Tm    Tm  \n",
       "0         NaN            9.5  60.0  \n",
       "1         NaN           15.9  60.0  \n",
       "2         NaN           17.4  60.0  \n",
       "3        32.7            NaN  60.0  \n",
       "4         NaN            NaN  60.0  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groupData['Geneious'] = reshapeGroupGeneious(groupData['Geneious'])\n",
    "groupData['Geneious'].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PrimerPair</th>\n",
       "      <th>Direction</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Start</th>\n",
       "      <th>Length</th>\n",
       "      <th>Tm</th>\n",
       "      <th>GC Percent</th>\n",
       "      <th>Amplicon</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>forward</td>\n",
       "      <td>GATGTGCCCAGATGGGATTAG</td>\n",
       "      <td>220.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>62.233</td>\n",
       "      <td>52.381</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>reverse</td>\n",
       "      <td>GTGGATGTCAAGACCAGGTAAG</td>\n",
       "      <td>994.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>62.205</td>\n",
       "      <td>50.000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2</td>\n",
       "      <td>forward</td>\n",
       "      <td>ACGGTAGCTAATACCGCATAAC</td>\n",
       "      <td>155.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>61.972</td>\n",
       "      <td>45.455</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>reverse</td>\n",
       "      <td>CCTCCAGATCTCTACGCATTTC</td>\n",
       "      <td>707.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>62.034</td>\n",
       "      <td>50.000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3</td>\n",
       "      <td>forward</td>\n",
       "      <td>TAGCGGTGAAATGCGTAGAG</td>\n",
       "      <td>679.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>61.797</td>\n",
       "      <td>50.000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   PrimerPair Direction                Sequence  Start  Length      Tm  \\\n",
       "0           1   forward   GATGTGCCCAGATGGGATTAG  220.0    21.0  62.233   \n",
       "1           1   reverse  GTGGATGTCAAGACCAGGTAAG  994.0    22.0  62.205   \n",
       "3           2   forward  ACGGTAGCTAATACCGCATAAC  155.0    22.0  61.972   \n",
       "4           2   reverse  CCTCCAGATCTCTACGCATTTC  707.0    22.0  62.034   \n",
       "6           3   forward    TAGCGGTGAAATGCGTAGAG  679.0    20.0  61.797   \n",
       "\n",
       "   GC Percent  Amplicon  \n",
       "0      52.381       NaN  \n",
       "1      50.000       NaN  \n",
       "3      45.455       NaN  \n",
       "4      50.000       NaN  \n",
       "6      50.000       NaN  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groupData['PrimerQuest'] = reshapeGroupPrimerQuest(groupData['PrimerQuest'])\n",
    "groupData['PrimerQuest'].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Initialize Species DataFrames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
